### DISCUSSION TITLE
Verifying SANS are present

### DISCUSSION THREAD

**Discussion by jax7778** [matteocorti/check_ssl_cert#291] [Q&A] [answered]
Hello, 

We recently upgraded our Nagios installation, and as part we updated the copy of this plugin. The changes made (it had been quite a while since it was last updated) seem to have broken our check. 

The goal of this check is to verify that the correct SANs are present on the current certificate. I know this is a bit of an odd check, but the last 3 times the cert has been replaced a crucial SAN has been left out, causing issues on our main site. The results of the current check appear to be that only the last SAN mentioned,  www.SAN4.com in this case, is the one checked. Any help is greatly appreciated. Thanks in advance. 

Here is an example with specific information removed: 
check_ssl_cert -H $HOSTADDRESS$ -p 443 -n *.SAN1.edu -n SAN1.edu -n SAN2.com -n www.SAN2.com -n SAN3.com -n www.SAN3.com -n SAN4.com -n www.SAN4.com --ignore-exp --altnames --format "%SHORTNAME% OK %DISPLAY_CN%%CHECKEDNAMES%"

**Comment by matteocorti** [comment]
Hi

There is a bug ...

I'll check

Matteo

**Comment by matteocorti** [comment, marked as answer]
Bug fixed in the latest release: can you please test?

**Reply by jax7778** [reply]
Thanks so much for getting back to me, and for looking into the issue. I didn't catch this until now, but I will check it on Monday when I am back in the office.

**Reply by jax7778** [reply]
Well, now we are getting an error that the hostname does not match the CN, which in a way is great.  In the command in the first post, the error reads. $HOSTADDRESS$ does not match  \*.SAN1.edu, adding in a backslash. If I remove *.SAN1.edu then the error clears, and the command appears to work as intended!

So the issue is mostly resolved. The only remaining issue is that it doesn't seem to be able to correctly interpret the wildcard cert, but other than that it appears to be working as intended. I thought the asterisk needed to be escape, but I tried the standard escape but it did not resolve the error. Any suggestions are greatly appreciated. 

Thanks so much for working on this issue.

**Reply by matteocorti** [reply]
Thanks. I reopened the issue. Do you have an example to test?

**Comment by matteocorti** [comment]
I am unable to reproduce the problem:

```
$  ./check_ssl_cert -H sp.ethz.ch -n '*.sp.ethz.ch' -n sherlock.sp.ethz.ch --ignore-exp --altnames --format "%SHORTNAME% OK %DISPLAY_CN%%CHECKEDNAMES%"
SSL_CERT OK '*.sp.ethz.ch' (sp.ethz.ch *.sp.ethz.ch sherlock.sp.ethz.ch) 
```

If you prefer not to publish an example, you can send it to me in private (although I don't really understand why a hostname with an HTTP server should be kept private). You can use: matteo@corti.li

**Reply by jax7778** [reply]
Sure, here is the command we are currently running: 
./check_ssl_cert -H www.uco.edu -p 443  -n *.uco.edu -n uco.edu -n endeavorgames.com -n www.endeavorgames.com -n ucojazzlab.com -n www.ucojazzlab.com --ignore-exp --altnames --format "%SHORTNAME% OK %DISPLAY_CN%%CHECKEDNAMES%"

**Reply by matteocorti** [reply]
Thanks for the example.

I'm unable to reproduce the problem:

```
$ ./check_ssl_cert -H www.uco.edu -p 443 -n *.uco.edu -n uco.edu -n endeavorgames.com -n www.endeavorgames.com -n ucojazzlab.com -n www.ucojazzlab.com --ignore-exp --altnames --format "%SHORTNAME% OK %DISPLAY_CN%%CHECKEDNAMES%"
SSL_CERT OK 'www.uco.edu' (www.uco.edu *.uco.edu uco.edu endeavorgames.com www.endeavorgames.com ucojazzlab.com www.ucojazzlab.com) 
```

Can you please execute the command with the debug option (```-d``` or ```--debug-file```) and post the output?

**Reply by jax7778** [reply]
Here: 

check_ssl_cert -H www.uco.edu -p 443 -n *.uco.edu -n uco.edu -n endeavorgames.com -n www.endeavorgames.com -n ucojazzlab.com -n www.ucojazzlab.com --ignore-exp --altnames --format "%SHORTNAME% OK %DISPLAY_CN%%CHECKEDNAMES%" -d
[DBG]OpenSSL 1.0.2k-fips 26 Jan 2017
[DBG]built on: reproducible build, date unspecified
[DBG]platform: linux-x86_64
[DBG]options: bn(64,64) md2(int) rc4(16x,int) des(idx,cisc,16,int) idea(int) blowfish(idx)
[DBG]compiler: gcc -I. -I.. -I../include -fPIC -DOPENSSL_PIC -DZLIB -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -DKRB5_MIT -m64 -DL_ENDIAN -Wall -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic -Wa,--noexecstack -DPURIFY -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
[DBG]OPENSSLDIR: "/etc/pki/tls"
[DBG]engines: rdrand dynamic
SSL_CERT CRITICAL www.uco.edu: invalid CN ('www.uco.edu' does not match '\*.uco.edu')

**Reply by matteocorti** [reply]
You are using a very old OpenSSL version which is not supported supported since 2019. You should upgrade.
Anyway: which distribution are you using (so that I can test on a VM)?

Did you post the complete output?

**Reply by lukastribus** [reply]
Looks like CentOS 7

**Reply by jax7778** [reply]
Nagios is running on RHEL 7. We should be able to get that OpenSSL version upgraded. That is the full output.

**Reply by matteocorti** [reply]
Ok. It's a RHEL backported version. I'll check on Centos 7.

**Reply by matteocorti** [reply]
I just checked on a RHEL 7 machine and cannot reproduce.

 * which version of check_ssl_cert are you running (```check_ssl_cert --version```)
 * the debugging output is very strange as several parts are missing, can you please try to generate it with the ```--debug-file``` option?

**Reply by jax7778** [reply]
Sorry for this late reply. 

Version: check_ssl_cert version 2.4.0

I noticed that you were on 2.4.1 . Updated to this version, same issue. 

I an attempt to get you the full output of the debug command, I ran the command manually from terminal, and it worked without issue. It is still failing in Nagios, but since it works from terminal I am not sure how to proceed. 

It keeps adding a backslash in front of the *.uco.edu address when ran from Nagios itself.  such as "\*.uco.edu" I can post the full output from the debug command from terminal, but since it shows successful, I am not sure how much assistance it would be. 

***Edit After messing around with quotes,  I got this to work. To get the command to pass properly I had to put two sets of single quotes around the wildcard SAN name like so ''*uco.edu''  does not work with two sets of double quotes. 

Hopefully that information helps. 

Thanks so much for all of your work on this.

**Reply by matteocorti** [reply]
This is a useful info.... Thanks!
I updated the README.md file
